torchtune/utils/_profiler.py

```diff
@@ -48,3 +61,405 @@ def trace_handler(prof) -> None:
+    output_dir,
+    metric="self_cuda_time_total",
+    row_limit=25,
+):
+    world_size, rank = get_world_size_and_rank()
+    curr_trace_dir_name = "iteration_" + str(prof.step_num)
+    curr_trace_dir = os.path.join(output_dir, curr_trace_dir_name)
+    if not os.path.exists(curr_trace_dir):
+        os.makedirs(curr_trace_dir, exist_ok=True)
+
+    # Export chrome / tensorboard trace
+    if rank == 0:
+        log.info(f"Dumping traces at step {prof.step_num}")
+    begin = time.monotonic()
+
+    # Use tensorboard trace handler rather than directly exporting chrome traces since
+    # tensorboard doesn't seem to be able to parse traces with prof.export_chrome_trace
+    exporter = tensorboard_trace_handler(
+        curr_trace_dir, worker_name=f"rank{rank}", use_gzip=True
+    )
+    exporter(prof)
+
+    if rank == 0:
+        log.info(f"Finished dumping traces in {time.monotonic() - begin:.2f} seconds")
+
+    # Memory timeline sometimes fails to export
+    if prof.profile_memory:
+        if rank == 0:
+            try:
+                prof.export_memory_timeline(
+                    f"{curr_trace_dir}/rank{rank}_memory-timeline.html"
+                )
+            except Exception as e:
+                log.warn(f" Failed to export memory timeline: {e}")
+
+    # Dump stack traces
+    if prof.with_stack:
+        prof.export_stacks(f"{curr_trace_dir}/rank{rank}_stacks.txt", metric=metric)
+
+    # Export event averages
+    key_avgs = prof.key_averages(
+        group_by_input_shape=prof.record_shapes, group_by_stack_n=5
+    ).table(sort_by=metric, row_limit=row_limit)
+    with open(f"{curr_trace_dir}/rank{rank}_key_averages.txt", "w") as f:
+        print(key_avgs, file=f)
+    if rank == 0:
+        log.info(f"Saving profiling results to {curr_trace_dir}")
+
+    # TODO: Is this necessary?
+    # see https://github.com/pytorch/torchtitan/blob/3050098dcee4901d88c712f9e8e9703d1735a29b/torchtitan/profiling.py#L48
+    if world_size > 1:
+        torch.distributed.barrier()
+
+
+class FakeProfiler:
+    """
+    Drop-in replacement for torch.profiler.profile that functions as a nullcontext / object
+    with no-op methods for `start`, `stop`, and `step`.
+
+    This is helpful for instrumenting profiling in a recipe without requiring changes to the
+    code independent of whether profiling is on / off.
+
+    E.g.,
+    ```
+        profiler = FakeProfiler()
+        #profiler = torch.profiler.profile()
+
+        # Below is same regardless of profiler object type
+        with profiler as prof:
+            for epoch in epochs:
+                for batch in batches:
+                    train.step()
+                    prof.step()
+    ```
+    """
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    def start(self):
+        pass
+
+    def stop(self):
+        pass
+
+    def step(self):
+        pass
+
+
+def should_profile(cfg: DictConfig) -> bool:
```

Comment:
Why not just check cfg.profiler.enabled directly and remove this util? This would reduce one level of indirection

Reply:
Done